### PR TITLE
Use `strip_{prefix,suffix}` in rustdoc

### PR DESCRIPTION
**Pull request by tesuji**
cc @Mark-Simulacrum @jyn514 for rustdoc part

Part of #74048

### CODE REVIEW

File: src/librustc_ast/util/comments.rs [hunk 1]
```diff
@@ -29,27 +29,31 @@ pub struct Comment {
...
-    let res = ((s.starts_with("/**") && *s.as_bytes().get(3).unwrap_or(&b' ') != b'*')
-        || s.starts_with("/*!"))
-        && s.len() >= 5;
-    debug!("is {:?} a doc comment? {}", s, res);
-    res
+    // Previously, `/**/` was incorrectly regarded as a doc comment because it
+    // starts with `/**` and ends with `*/`. However, this caused an ICE
+    // because some code assumed that the length of a doc comment is at least 5.
+    let yes = match s.as_bytes() {
+        [b'/', b'*', b'*', c, _, ..] => *c != b'*',
```
Comment (GuillaumeGomez) [inline src/librustc_ast/util/comments.rs:47]: I'm surprised that we still allow `/**` to be a doc comment. Is it something allowed officially?

**Comment by GuillaumeGomez** [comment]
Also, what case is it fixing? Because if it's fixing anything, please add a test. :)

File: src/librustc_ast/util/comments.rs [hunk 2]
```diff
@@ -29,27 +29,31 @@ pub struct Comment {
...
 }
 
-// FIXME(#64197): Try to privatize this again.
-pub fn is_doc_comment(s: &str) -> bool {
-    (s.starts_with("///") && is_line_doc_comment(s))
-        || s.starts_with("//!")
-        || (s.starts_with("/**") && is_block_doc_comment(s))
-        || s.starts_with("/*!")
+fn is_doc_comment(s: &str) -> bool {
+    is_line_doc_comment(s) || is_block_doc_comment(s)
```
Comment (Mark-Simulacrum) [inline src/librustc_ast/util/comments.rs:56]: I left a review on the previous PR here - this is a change in behavior due to checking the length for e.g. `/*!`. I would ask that any changes in behavior are at least split into their own commit in this PR and as such clearly identifiable, ideally, they would be made in a separate PR. This one for example seems to have backwards compatibility concerns.

**Review by Mark-Simulacrum — CHANGES_REQUESTED** [review]
I am concerned that it seems like at least some of my comments from the previous PR have not been addressed. I would prefer not to have to review again or try to copy my comments over myself...

File: src/librustc_ast/util/comments.rs (hunk 1 above)
Comment (lzutao) [reply src/librustc_ast/util/comments.rs:47]: Here is one: https://github.com/facebook/relay/blob/07ccab7cc637f51f2f15fc75ed824d1de8ede72f/compiler/crates/schema/src/parser.rs#L23-L26

File: src/librustc_ast/util/comments.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lzutao) [reply src/librustc_ast/util/comments.rs:56]: Also there are no testcases cover this function?

File: src/librustc_ast/util/comments.rs [hunk 3]
```diff
@@ -127,22 +131,26 @@ pub fn strip_doc_comment_decoration(comment: &str) -> String {
...
         }
     }
 
-    if comment.starts_with("/*") {
-        let lines =
-            comment[3..comment.len() - 2].lines().map(|s| s.to_string()).collect::<Vec<String>>();
+    match comment
+        .strip_prefix("/**")
+        .or_else(|| comment.strip_prefix("/*!"))
+        .and_then(|s| s.strip_suffix("*/"))
```
Comment (lzutao) [inline src/librustc_ast/util/comments.rs:142]: re https://github.com/rust-lang/rust/pull/74048/files#r449877722

I am quite sure that with block doc comment must start with either `/**` or `/*!` and must end with `*/`
as the reference states: https://doc.rust-lang.org/nightly/reference/comments.html#doc-comments

File: src/librustdoc/clean/utils.rs [hunk 4]
```diff
@@ -481,12 +481,14 @@ pub fn print_const(cx: &DocContext<'_>, n: &'tcx ty::Const<'_>) -> String {
...
-                    let n = s.len() - ": ".len();
-                    s.truncate(n);
+            if let Some(head) = s.strip_suffix("_usize") {
+                let new_len = match head.strip_suffix(": ") {
+                    None => head.len(),
+                    Some(hhead) => hhead.len(),
+                };
+                // SAFETY: `new_len` should be in between char boundary
+                unsafe {
+                    s.as_mut_vec().set_len(new_len);
```
Comment (lzutao) [inline src/librustdoc/clean/utils.rs:491]: re https://github.com/rust-lang/rust/pull/74048/files#r449880719

`String::truncate` is cheap but there is panicking branch not elided: https://rust.godbolt.org/z/VTZ3Le
Also miri confirms that this is safe.

File: src/librustdoc/html/markdown.rs [hunk 5]
```diff
@@ -126,26 +145,6 @@ impl<'a> Line<'a> {
     }
 }
 
-// FIXME: There is a minor inconsistency here. For lines that start with ##, we
-// have no easy way of removing a potential single space after the hashes, which
-// is done in the single # case. This inconsistency seems okay, if non-ideal. In
-// order to fix it we'd have to iterate to find the first non-# character, and
-// then reallocate to remove it; which would make us return a String.
```
Comment (lzutao) [inline src/librustdoc/html/markdown.rs:133]: re https://github.com/rust-lang/rust/pull/74048/files#r449881015

I have no idea what this comment is talking about.
Does it replace "##" with "#" as the old code did?
So should I keep the old FIXME?

Comment (lzutao) [reply src/librustdoc/html/markdown.rs:133]: Oh you wrote this code: ffe12b1274732e791e117d8cc79b3db0f4161956 and has a test here: https://github.com/rust-lang/rust/blob/master/src/test/rustdoc/issue-41783.rs

File: src/librustdoc/html/markdown.rs [hunk 6]
```diff
@@ -776,10 +775,10 @@ impl LangString {
                     data.no_run = true;
                 }
                 x if x.starts_with("edition") => {
-                    data.edition = x[7..].parse::<Edition>().ok();
+                    data.edition = x.strip_prefix("edition").unwrap().parse::<Edition>().ok();
```
Comment (lzutao) [inline src/librustdoc/html/markdown.rs:778]: re https://github.com/rust-lang/rust/pull/74048/files#r449881344

I changed the code to use `strip_prefix` to avoid panicking code caused by slicing.

File: src/librustdoc/html/render.rs [hunk 7]
```diff
@@ -810,11 +810,9 @@ themePicker.onblur = handleThemeButtonsBlur;
                 if line.starts_with(&format!("\"{}\"", krate)) {
                     continue;
                 }
-                if line.ends_with(",\\") {
-                    ret.push(line[..line.len() - 2].to_string());
-                } else {
+                if let Some(head) = line.strip_suffix(",\\").or_else(|| line.strip_suffix("\\")) {
                     // Ends with "\\" (it's the case for the last added crate line)
-                    ret.push(line[..line.len() - 1].to_string());
+                    ret.push(head.to_string());
```
Comment (lzutao) [inline src/librustdoc/html/render.rs:815]: re https://github.com/rust-lang/rust/pull/74048/files#r449881783

The comment above this statement confirms that the `else` case `line` ends with "\\".
cc @GuillaumeGomez who wrote this code in b4fb3069ce82f61f84a9487d17fb96389d55126a

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 8]
```diff
@@ -581,28 +574,27 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
...
-                    link.trim_end_matches("()")
-                } else if link.starts_with("macro@") {
-                    kind = Some(MacroNS);
-                    link.trim_start_matches("macro@")
-                } else if link.starts_with("derive@") {
-                    kind = Some(MacroNS);
-                    link.trim_start_matches("derive@")
-                } else if link.ends_with('!') {
-                    kind = Some(MacroNS);
-                    link.trim_end_matches('!')
```
Comment (lzutao) [inline src/librustdoc/passes/collect_intra_doc_links.rs:601]: re https://github.com/rust-lang/rust/pull/74048/files#r449882381

Quoting [intra doc link RFC](https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md#path-ambiguities):

> - In links to macros,
>  the link label can end with a `!`,
>  e.g., `Look at the [FOO!] macro`. You can alternatively use a `macro@` prefix,
>  e.g. `[macro@foo]`

I don't know if multiple `!` is allowed in that RFC. cc @Manishearth

File: src/librustdoc/html/render.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render.rs]: Normally, all lines should end with ",\\" (except the last one which ends with "\\").

Comment (lzutao) [reply src/librustdoc/html/render.rs]: So does I reserve old behavior here?

**Comment by lzutao** [comment]
I replied to most review comments about rustdoc (#74048).

File: src/librustdoc/html/render.rs [hunk 9]
```diff
@@ -810,11 +810,9 @@ themePicker.onblur = handleThemeButtonsBlur;
                 if line.starts_with(&format!("\"{}\"", krate)) {
                     continue;
                 }
-                if line.ends_with(",\\") {
-                    ret.push(line[..line.len() - 2].to_string());
-                } else {
-                    // Ends with "\\" (it's the case for the last added crate line)
-                    ret.push(line[..line.len() - 1].to_string());
+                // Ends with "\\" (it's the case for the last added crate line)
+                if let Some(head) = line.strip_suffix(",\\").or_else(|| line.strip_suffix("\\")) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render.rs:814]: Like this seems fine to me, thanks!

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 8 above)
Comment (Manishearth) [reply src/librustdoc/passes/collect_intra_doc_links.rs:601]: Only a single one

File: src/librustc_ast/util/comments.rs [hunk 10]
```diff
@@ -29,27 +29,31 @@ pub struct Comment {
...
+    let yes = match s.as_bytes() {
+        [b'/', b'*', b'*', c, _, ..] => *c != b'*',
+        [b'/', b'*', b'!', _, _, ..] => true,
+        _ => false,
+    };
+    debug!("is {:?} a block doc comment? {}", s, yes);
+    yes
 }
 
-// FIXME(#64197): Try to privatize this again.
```
[the remark this answers is not in the document]
Comment (Mark-Simulacrum) [reply src/librustc_ast/util/comments.rs:47]: I continue to prefer that we not conflate the move to using strip_{prefix,suffix} with other changes, just because it makes it harder to review and increases the chances of missing something. At the very least combining them in the same commit is quite frustrating for things like git blame later on.

**Comment by Mark-Simulacrum** [comment]
I continue to be pretty strongly opposed to making both functional changes, regardless of whether we think they're fine, and "pure cleanup" of using the new functions from std in the same PR. At the very least, I am not confident that my original review caught all such cases -- and without separate commits at least for them, it makes it hard to evaluate whether the breaking changes being made are acceptable or not.

I will try to reply to your comments shortly.

**Comment by lzutao** [comment]
Many behavior changes are either:
* replacing `trim_*_matches` with `strip_*`?
* make it clear what we strip: https://github.com/rust-lang/rust/pull/74094#discussion_r450251662

The first case, most of them I think the intent is really clear as there is an if condition check if str starts/ends with pattern only once.
The second case is needed if we want to use `str::strip_*` methods and leverage its usefulness.

**Comment by Mark-Simulacrum** [comment]
I don't disagree that it's likely that the behavior changes are benign, but I am personally against landing a PR that has a bunch of (very minor) breaking changes to various parts of the codebase without any evaluation of what those changes are exactly. I also don't really trust myself to have caught every case where there *has* been a behavior change, especially when some of those are being made intentionally it seems (so it's not just "accidents" that review needs to catch).

I think it's generally a good idea to have a breaking change "set" at least split out into its own PR and then we can evaluate whether the code readability of strip_prefix is worth potentially breaking users, even if we unintentionally were accepting behavior of some kind before. Making that evaluation without a list of the changes being made is quite difficult.

File: src/librustdoc/html/render.rs (hunk 7 above)
Comment (Manishearth) [reply src/librustdoc/html/render.rs]: Yes please

**Comment by Manishearth** [comment]
I agree with @Mark-Simulacrum , we should not mix behavior changes with refactors like this.

The changes being benign is already false: you introduced a bug in https://github.com/rust-lang/rust/pull/74094#discussion_r450272344

File: src/librustc_ast/util/comments.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustc_ast/util/comments.rs:47]: TIL. :)

**Comment by Muirrum** [comment]
@lzutao This is a triage bump

**Comment by lzutao** [comment]
This is blocked on other PR. Closing for now!
